### ISSUE TITLE
Is it possible to generate a DIFF / patch for Blender 2.80 with the changes that were done in this repo?

### ISSUE THREAD

**Issue by aliasguru** [MARUI-PlugIn/BlenderXR#15]
I'm asking this because I'm compiling Blender 2.80 myself, while adding little things on top of it. One thing that I'd like to add is the ability to open a VR window. For this, I'd need the changes that were done by you to the Blender master branch as a patch / DIFF. I am not talking about the MarUI plugin, I am only talking about the changes you had to make to Blender itself.

**Comment by pk15950** [comment]
Hi aliasguru,
As of now, generating a diff would be difficult given that the code in this repo is based off the state of the blender2.8 branch (which I believe is now master) of Blender taken on January 10, 2019. In the future, we'll make sure to generate a diff every time we merge with upstream Blender.

However, to keep track of our changes, we actually have a preprocessor directive that we include everywhere that we alter original Blender code. If you search all files for "#if WITH_VR", you'll be able to find all of our changes.

In addition to that, there are a few other important changes that we had to make to add a VR window:
1. release/scripts/startup/bl_ui/space_topbar.py
Added operators for the VR window to the python layout file (wm.window_new_vr).
2. wm_operators.c
Added operators for the VR window (WM_OT_window_new_vr).   
3. wm_window.c / wm_window.h
Implemented the actual VR window (wm_window_new_vr_exec)
 
Finally, the entire bf_vr module (source/blender/vr) consists of code we added. The shared library loading and "attaching" of the headset and window to the viewport is handled in vr_main.c. 
In order to add this extra module to Blender, entries need to be added to the following files:
1. source/blender/CMakeLists.txt 
2. build_files/cmake/macros.cmake

Hope this helps!

**Comment by aliasguru** [comment]
Awesome, thanks for the reply. I'll check and see how far I can get. I'll leave the ticket open for a while in case I run into issues.

**Comment by pk15950** [comment]
Here's the diff for the latest "release": 
https://github.com/MARUI-PlugIn/BlenderXR/blob/master/BlenderXR-diff-20190208.patch

**Comment by aliasguru** [comment]
Awesome! The patch applied fine on latest master. Do you plan to update it in the future as well?

**Comment by pk15950** [comment]
Good to hear! Yes, the patch will be updated every time we merge upstream.

**Comment by aliasguru** [comment]
I have successfully adapted your Patch to work with today's master. Only one thing I don't really get to work. In `vr_widget_separate.cpp`, there is a code block around line 105, which tries to copy custom data:

```
	CustomData_copy(&bm_old->vdata, &bm_new->vdata, CD_MASK_BMESH, CD_CALLOC, 0);
	CustomData_copy(&bm_old->edata, &bm_new->edata, CD_MASK_BMESH, CD_CALLOC, 0);
	CustomData_copy(&bm_old->ldata, &bm_new->ldata, CD_MASK_BMESH, CD_CALLOC, 0);
	CustomData_copy(&bm_old->pdata, &bm_new->pdata, CD_MASK_BMESH, CD_CALLOC, 0);
```

However, this block does not compile (I simply commented it out which makes BlenderXR compile successfully, and it works on the Vive). The error in VS Code and the compilation console is:

> no suitable conversion function from "const CustomData_MeshMasks" to "CustomDataMask" exists

It seems like the `CustomData_copy()` method at some point has changed what has to be passed on as input. `CD_MASK_BMESH` is of type `CustomData_MeshMasks`, but the method requires `CustomDataMask` as type.

Any ideas? If you want, I can provide the patch back to you if you see it useful, I can transfer it via Google Drive.

**Comment by pk15950** [comment]
Thanks for the feedback, and glad to hear that you were able to make it work with today's master.

The thing is, the BlenderXR repo was never meant to be used as a patch for later Blender developments. Rather, it was meant to be sort of a "snapshot" of using VR with Blender at the time the repo is updated (the last update being February 13th). 

The reason behind this strategy is Blender is constantly being changed and refactored, so it would be difficult to ensure that Blender functions used by the VR module continue to work properly in the future.

This seems to be the case with the compilation error you were getting. Most likely, since February 13th, the CustomData_copy() method has indeed changed so it will require different inputs. 

I'm not exactly sure what the quick fix is, but you might be able to copy-paste the "mesh_separate_tagged()" method in editmesh_tools.c and replace the one in vr_widget_separate.cpp since the functionality is meant to be the same anyways. 

The strategy for updating the repo might change in the future, but otherwise you will have to fix any compilation errors manually when merging upstream.

**Comment by aliasguru** [comment]
You were right, that code block copied over from `mesh_separate_tagged()` worked just fine, the code compiles again! Thanks so much for the effort. I've updated the patch on Google Drive too.

I do understand that you're following the concept of having snapshots, and not constantly rebase against master. While adapting the patch I've seen in how many locations simple API renaming broke the patch. The reason why I went through this is that we also use custom Python tools for modelling here in the studio. Those tools break just as fast, because in the Python API there is also still massive renaming going on. So I ran into the situation that current master and BlenderXR were not compatible any longer in regards to Python, and my Add-ons stopped working!

Now thanks to your help I basically have one Blender version only that needs to be maintained: BlenderXR  :)
